downsampler/src/main/scala/filodb.downsampler/DownsamplerMain.scala

```diff
@@ -0,0 +1,108 @@
+
+
+  /**
+    * Downsample batch of raw partitions, and store downsampled chunks to cassandra
+    */
+  private[downsampler] def downsampleBatch(rawPartsBatch: Seq[PartDataRow]) = {
+
+    logger.debug(s"Starting downsampling job for rawDataset=$rawDatasetName for " +
+      s"ingestionTimeStart=$ingestionTimeStart ingestionTimeEnd=$ingestionTimeEnd " +
+      s"userTimeStart=$userTimeStart userTimeEnd=$userTimeEnd for ${rawPartsBatch.size} partitions")
+
+    val downsampledChunksToPersist = MMap[FiniteDuration, Iterator[ChunkSet]]()
+    downsampleResolutions.foreach { res =>
+      downsampledChunksToPersist(res) = Iterator.empty
+    }
+    val rawPartsToFree = ArrayBuffer[PagedReadablePartition]()
+    val downsampledPartsPartsToFree = ArrayBuffer[TimeSeriesPartition]()
+
+    rawPartsBatch.foreach { rawPart =>
+
+      val rawSchemaId = RecordSchema.schemaID(rawPart.partData.partitionKey, UnsafeUtils.arayOffset)
+      val rawPartSchema = schemas(rawSchemaId)
+      val downsampleSchema = rawPartSchema.downsample.get
+      logger.debug(s"Downsampling partition ${rawPartSchema.partKeySchema.stringify(rawPart.partData.partitionKey)} ")
+
+      val rawReadablePart = new PagedReadablePartition(rawPartSchema, 0, 0,
+        rawPart.partData, memoryManager)
+      val downsampledParts = PartitionDownsampler.downsample(shardStats,
+        downsampleResolutions, downsampleSchema, chunkDownsamplersByRawSchemaId(rawSchemaId),
+        rawReadablePart, blockFactory, bufferPoolByRawSchemaId(rawSchemaId),
+        memoryManager, userTimeStart, // use userTime as ingestionTime for downsampled data
+        userTimeStart, userTimeEnd)
+      rawPartsToFree += rawReadablePart
+      downsampledPartsPartsToFree ++= downsampledParts.values
+
+      downsampledParts.foreach { case (res, dsPartition) =>
+        dsPartition.switchBuffers(blockFactory, true)
+        downsampledChunksToPersist(res) ++= dsPartition.makeFlushChunks(blockFactory)
+      }
+    }
+    persistDownsampledChunks(downsampledChunksToPersist)
+
+    // reclaim all blocks
+    blockFactory.markUsedBlocksReclaimable()
+    // free partitions
+    rawPartsToFree.foreach(_.free())
+    rawPartsToFree.clear()
+    downsampledPartsPartsToFree.foreach(_.shutdown())
+    downsampledPartsPartsToFree.clear()
+
+    logger.info(s"Finished iterating through and downsampling ${rawPartsBatch.size} partitions in current executor")
+  }
+
+  /**
+    * Persist chunks in `downsampledChunksToPersist`
+    */
+  private def persistDownsampledChunks(downsampledChunksToPersist: MMap[FiniteDuration, Iterator[ChunkSet]]): Unit = {
+    // write all chunks to cassandra
+    val writeFut = downsampledChunksToPersist.map { case (res, chunks) =>
+      PerSparkExecutorState.cassandraColStore.write(downsampleDatasets(res).ref,
+        Observable.fromIterator(chunks), ttlByResolution(res))
+    }
+
+    writeFut.foreach { fut =>
+      val response = Await.result(fut, cassWriteTimeout)
```

Comment:
Do we want writes to be serial?  There aren't many resolutions, it seems fine to write them all at once.

Also, since this is a bulk write, I think it's better to use the Spark Cassandra Connector to do it.

Reply:
The parallelism is achieved by the cassandra splits.

I'll certainly look into the bulk-writer here. We can improve iteratively.